### PR TITLE
feat: use update_transform when possible

### PR DESCRIPTION
**Pull request by thebrianchen**

### CODE REVIEW

File: dev/src/write-batch.ts [hunk 1]
```diff
@@ -479,13 +479,26 @@ export class WriteBatch {
...
         write = document.toProto();
         write!.updateMask = documentMask.toProto();
-      }
 
-      return {
-        write,
-        transform: transform.toProto(this._serializer),
-        precondition: precondition.toProto(),
-      };
+        const transformProto = transform.toProto(this._serializer);
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: Can we update `transform.toProto` to return `transformProto.transform.fieldTransforms`? Ideally we could just write:

```
write = document.toProto();
write!.updateMask = documentMask.toProto();	        
write!. updateTransforms = transform.toProto();
```

```

File: dev/src/write-batch.ts [hunk 2]
```diff
@@ -479,13 +479,26 @@ export class WriteBatch {
...
+        return {
+          write,
+          precondition: precondition.toProto(),
+        };
+      } else {
+        return {
+          write,
+          transform: transform.toProto(this._serializer),
+          precondition: precondition.toProto(),
+        };
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: Is it possible to completely replace our usage of `transform` with `write.updateTransform`?

File: dev/src/write-batch.ts (hunk 1 above)
Comment (thebrianchen) [reply dev/src/write-batch.ts]: It's currently being used that way in `set` and `create`, so I didn't change it. I could also write it as a separate method like `toFieldTransformProto`, but that seems like extra code. Thoughts?

Comment (schmidt-sebastian) [reply dev/src/write-batch.ts]: You should replace all usages of `transform` with `write. updateTransforms`. If you don't, you will get super confused later on. Trust me :)

Comment (thebrianchen) [reply dev/src/write-batch.ts]: done. grasping for the light

File: dev/src/write-batch.ts (hunk 2 above)
Comment (thebrianchen) [reply dev/src/write-batch.ts]: i think so!

File: dev/system-test/firestore.ts [hunk 3]
```diff
@@ -1975,19 +1975,26 @@ describe('Transaction class', () => {
   it('has update() method', () => {
     const ref = randomCol.doc('doc');
     return ref
-      .set({foo: 'bar'})
+      .set({
+        foo: ['bar', 'baz'],
```
Comment (schmidt-sebastian) [inline dev/system-test/firestore.ts]: Nit: Either all foo/bar/baz or farm animals please :)

File: dev/system-test/firestore.ts [hunk 4]
```diff
@@ -1975,19 +1975,26 @@ describe('Transaction class', () => {
...
+            });
           });
         });
       })
       .then(() => {
         return ref.get();
       })
       .then(doc => {
-        expect(doc.get('foo')).to.equal('foobar');
+        expect(doc.get('foo')).to.deep.equal(['baz']);
```
Comment (schmidt-sebastian) [inline dev/system-test/firestore.ts]: Nit: Use doc.data() and simplify into one line.

File: dev/test/util/helpers.ts [hunk 5]
```diff
@@ -89,27 +89,25 @@ export function verifyInstance(firestore: Firestore): Promise<void> {
...
-    const update = Object.assign({}, document);
-    delete update.updateTime;
-    delete update.createTime;
-    writes.push({update});
     if (mask) {
       writes[0].updateMask = mask;
     }
   }
 
   if (transforms) {
```
Comment (schmidt-sebastian) [inline dev/test/util/helpers.ts]: Same question - could we just remove these if statements?

File: dev/test/util/helpers.ts [hunk 6]
```diff
@@ -124,47 +122,43 @@ export function updateMask(...fieldPaths: string[]): api.IDocumentMask {
...
   transforms?: api.DocumentTransform.IFieldTransform[];
   mask?: api.IDocumentMask;
 }): api.ICommitRequest {
   return write(
-    opts.document || null,
+    opts.document,
     opts.mask || null,
     opts.transforms || null,
-    null
+    /** precondition */ null
```
Comment (schmidt-sebastian) [inline dev/test/util/helpers.ts]: Nit: Missing equals

File: dev/test/util/helpers.ts [hunk 7]
```diff
@@ -124,47 +122,43 @@ export function updateMask(...fieldPaths: string[]): api.IDocumentMask {
...
   precondition?: api.IPrecondition;
 }): api.ICommitRequest {
   const precondition = opts.precondition || {exists: true};
   const mask = opts.mask || updateMask();
   return write(
-    opts.document || null,
+    opts.document,
     mask,
     opts.transforms || null,
+    // opts.updateTransforms || null,
```
Comment (schmidt-sebastian) [inline dev/test/util/helpers.ts]: Please fix.

File: dev/src/write-batch.ts [hunk 8]
```diff
@@ -194,12 +193,13 @@ export class WriteBatch {
 
     const op = () => {
       const document = DocumentSnapshot.fromObject(documentRef, firestoreData);
-      const write =
-        !document.isEmpty || transform.isEmpty ? document.toProto() : null;
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: Can you also check if we still use `document.isEmpty`?

File: dev/test/util/helpers.ts (hunk 5 above)
Comment (thebrianchen) [reply dev/test/util/helpers.ts]: We can't remove them unless we want to also remove the checks in the tests and always include the protos even if they are null/undefined.

File: dev/src/write-batch.ts (hunk 8 above)
Comment (thebrianchen) [reply dev/src/write-batch.ts]: removed.

File: dev/src/document.ts [hunk 9]
```diff
@@ -971,29 +961,19 @@ export class DocumentTransform<T = DocumentData> {
...
       fieldTransforms.push(transform.toProto(serializer, path));
     }
 
-    return {
-      transform: {
-        document: this.ref.formattedName,
-        fieldTransforms,
-      },
-    };
+    return fieldTransforms;
```
Comment (schmidt-sebastian) [inline dev/src/document.ts]: Optional: Use `map` to turn this function into one line.

File: dev/src/write-batch.ts [hunk 10]
```diff
@@ -324,20 +324,18 @@ export class WriteBatch {
...
-        write = document.toProto();
-      } else if (hasDocumentData || transform.isEmpty) {
-        write = document.toProto()!;
+      const hasMerge = mergePaths || mergeLeaves;
+      if (hasMerge) {
         write.updateMask = documentMask!.toProto();
       }
 
       return {
         write,
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: Nit: remove trailing comma.

File: dev/src/write-batch.ts [hunk 11]
```diff
@@ -557,22 +552,15 @@ export class WriteBatch {
     request.writes = [];
 
     for (const req of writes) {
-      assert(
-        req.write || req.transform,
-        'Either a write or transform must be set'
-      );
+      assert(req.write, 'A write must be set');
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/write-batch.ts]: Can we not make `write` non-optional now?

File: dev/src/write-batch.ts [hunk 12]
```diff
@@ -557,22 +552,15 @@ export class WriteBatch {
...
 
       if (req.precondition) {
-        (req.write || req.transform)!.currentDocument = req.precondition;
+        req.write!.currentDocument = req.precondition;
```
[the remark this answers is not in the document]
Comment (schmidt-sebastian) [reply dev/src/write-batch.ts]: Sounds good. Let's leave it as is.

File: dev/src/write-batch.ts [hunk 13]
```diff
@@ -602,23 +588,7 @@ export class WriteBatch {
...
-            if (writeRequest.write && writeRequest.transform) {
-              // The document transform is always sent last and produces the
-              // latest update time.
-              ++offset;
-            }
-
-            const writeResult = resp.writeResults[i + offset];
-
+          for (const writeResult of resp.writeResults) {
             writeResults.push(
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: This could also just be a single `map` call now.

File: dev/system-test/firestore.ts [hunk 14]
```diff
@@ -1975,19 +1975,28 @@ describe('Transaction class', () => {
...
+      .set({
+        boo: ['ghost', 'sebastian'],
+        moo: 'chicken',
+      })
       .then(() => {
         return firestore.runTransaction(updateFunction => {
           return updateFunction.get(ref).then(() => {
-            updateFunction.update(ref, {foo: 'foobar'});
+            updateFunction.update(ref, {
+              boo: FieldValue.arrayRemove('sebastian'),
```
Comment (schmidt-sebastian) [inline dev/system-test/firestore.ts]: So sad. Tears.

:)

File: dev/test/document.ts [hunk 15]
```diff
@@ -913,6 +913,8 @@ describe('set document', () => {
         requestEquals(
           request,
           set({
+            document: document('documentId'),
+            mask: updateMask(),
```
Comment (schmidt-sebastian) [inline dev/test/document.ts]: This will break the conformance tests. We should also not set an empty update mask.

File: dev/test/util/helpers.ts [hunk 16]
```diff
@@ -124,47 +120,37 @@ export function updateMask(...fieldPaths: string[]): api.IDocumentMask {
...
 }): api.ICommitRequest {
-  return write(
-    opts.document || null,
-    /* updateMask */ null,
-    opts.transforms || null,
-    {
-      exists: false,
-    }
-  );
+  return write(opts.document, /* updateMask */ null, opts.transforms || null, {
```
Comment (schmidt-sebastian) [inline dev/test/util/helpers.ts]: Missing `=` in comment.

File: dev/src/write-batch.ts (hunk 11 above)
[the remark this answers is not in the document]
Comment (thebrianchen) [reply dev/src/write-batch.ts]: you are write

File: dev/src/document.ts (hunk 9 above)
Comment (thebrianchen) [reply dev/src/document.ts]: attempted, but used `Array.from()`

File: dev/test/document.ts (hunk 15 above)
Comment (thebrianchen) [reply dev/test/document.ts]: removed here and in other unit tests that used an empty update mask

File: dev/src/write-batch.ts [hunk 17]
```diff
@@ -557,22 +552,11 @@ export class WriteBatch {
...
-
-      if (req.write) {
-        request.writes.push(req.write);
+        req.write!.currentDocument = req.precondition;
       }
 
-      if (req.transform) {
-        request.writes.push(req.transform);
-      }
+      request.writes.push(req.write!);
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: Here and above: Remove bang operator after `write`.

File: dev/src/write-batch.ts [hunk 18]
```diff
@@ -589,7 +573,7 @@ export class WriteBatch {
     return this._firestore
       .request<api.ICommitRequest, api.CommitResponse>('commit', request, tag)
       .then(resp => {
-        const writeResults: WriteResult[] = [];
+        let writeResults: WriteResult[] = [];
 
         if (request.writes!.length > 0) {
           assert(
```
Comment (schmidt-sebastian) [inline dev/src/write-batch.ts]: This assert is not needed anymore - we no longer use index-based lookups in this map. That means that you can remove the assert and the intermediate writeResult variable.

File: dev/test/util/helpers.ts [hunk 19]
```diff
@@ -124,47 +120,37 @@ export function updateMask(...fieldPaths: string[]): api.IDocumentMask {
...
   transforms?: api.DocumentTransform.IFieldTransform[];
   mask?: api.IDocumentMask;
 }): api.ICommitRequest {
   return write(
-    opts.document || null,
+    opts.document,
     opts.mask || null,
     opts.transforms || null,
-    null
+    /** precondition= */ null
```
Comment (schmidt-sebastian) [inline dev/test/util/helpers.ts]: Nit: Only one asterisk to start.
